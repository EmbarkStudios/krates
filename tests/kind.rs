mod util;

use util::build;

#[test]
fn all_the_things() {
    let grafs = build("all-features.json", krates::Builder::new()).unwrap();

    assert!(grafs.filtered.is_empty());
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        // This shouldn't actually affect anything, as dev dependencies
        // for non-workspace crates are already not resolved
        assert!(grafs.filtered.is_empty());
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        // This will be equivalent to to filtering workspace dev crates
        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_normal() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build_and_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn only_b() {
    let mut kb = krates::Builder::new();
    kb.include_workspace_crates(["/home/jake/code/krates/tests/ws/b/Cargo.toml"]);

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_after_build() {
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Dev);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }

    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Build);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }
}
