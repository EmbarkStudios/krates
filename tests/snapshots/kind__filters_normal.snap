---
source: tests/kind.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 (path+file:///home/jake/code/krates/tests/ws/a)" ]
    1 [ label = "crate b 0.1.0 (path+file:///home/jake/code/krates/tests/ws/b)" ]
    2 [ label = "crate c 0.1.0 (path+file:///home/jake/code/krates/tests/ws/c)" ]
    3 [ label = "crate cc 1.0.73 (git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6)" ]
    4 [ label = "crate cc 1.0.73 (registry+https://github.com/rust-lang/crates.io-index)" ]
    5 [ label = "crate coreaudio-rs 0.9.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    6 [ label = "crate difference 2.0.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    7 [ label = "crate lazy_static 1.4.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    8 [ label = "crate leftpad 0.2.0 (registry+https://github.com/rust-lang/crates.io-index)" ]
    9 [ label = "crate libc 0.2.133 (registry+https://github.com/rust-lang/crates.io-index)" ]
    10 [ label = "crate nix 0.16.1 (registry+https://github.com/rust-lang/crates.io-index)" ]
    11 [ label = "crate ring 0.16.20 (registry+https://github.com/rust-lang/crates.io-index)" ]
    12 [ label = "crate spin 0.5.2 (registry+https://github.com/rust-lang/crates.io-index)" ]
    13 [ label = "crate wasm-bindgen-futures 0.4.33 (registry+https://github.com/rust-lang/crates.io-index)" ]
    14 [ label = "crate web-sys 0.3.60 (registry+https://github.com/rust-lang/crates.io-index)" ]
    15 [ label = "crate winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)" ]
    16 [ label = "feature default" ]
    17 [ label = "feature default" ]
    18 [ label = "feature default" ]
    19 [ label = "feature default" ]
    20 [ label = "feature Crypto" ]
    21 [ label = "feature Window" ]
    22 [ label = "feature leftpad" ]
    23 [ label = "feature leftier-strings" ]
    24 [ label = "feature lazy_static" ]
    25 [ label = "feature audio_toolbox" ]
    26 [ label = "feature audio_unit" ]
    27 [ label = "feature core_audio" ]
    28 [ label = "feature open_al" ]
    29 [ label = "feature core_midi" ]
    30 [ label = "feature alloc" ]
    31 [ label = "feature dev_urandom_fallback" ]
    32 [ label = "feature once_cell" ]
    33 [ label = "feature EventTarget" ]
    0 -> 1 [ label = "" ]
    0 -> 16 [ label = "(dev)" ]
    0 -> 16 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 16 [ label = "" ]
    1 -> 17 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    1 -> 13 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    1 -> 13 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    2 -> 3 [ label = "(build)" ]
    2 -> 18 [ label = " 'x86_64-apple-darwin'" ]
    2 -> 19 [ label = "(dev)" ]
    2 -> 7 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    2 -> 8 [ label = "" ]
    2 -> 9 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    2 -> 10 [ label = " 'x86_64-unknown-linux-gnu'" ]
    2 -> 12 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    2 -> 20 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    2 -> 21 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    2 -> 15 [ label = " 'cfg(target_os = \"windows\")'" ]
    10 -> 4 [ label = "(build) 'cfg(target_os = \"dragonfly\")'" ]
    11 -> 4 [ label = "(build)" ]
    22 -> 2 [ label = "" ]
    22 -> 8 [ label = "" ]
    23 -> 2 [ label = "" ]
    23 -> 22 [ label = "" ]
    24 -> 2 [ label = "" ]
    24 -> 7 [ label = "" ]
    16 -> 2 [ label = "" ]
    16 -> 22 [ label = "" ]
    18 -> 5 [ label = "" ]
    18 -> 25 [ label = "" ]
    18 -> 26 [ label = "" ]
    18 -> 27 [ label = "" ]
    18 -> 28 [ label = "" ]
    18 -> 29 [ label = "" ]
    29 -> 5 [ label = "" ]
    28 -> 5 [ label = "" ]
    27 -> 5 [ label = "" ]
    26 -> 5 [ label = "" ]
    25 -> 5 [ label = "" ]
    19 -> 6 [ label = "" ]
    17 -> 11 [ label = "" ]
    17 -> 30 [ label = "" ]
    17 -> 31 [ label = "" ]
    31 -> 11 [ label = "" ]
    31 -> 32 [ label = "" ]
    32 -> 11 [ label = "" ]
    30 -> 11 [ label = "" ]
    21 -> 14 [ label = "" ]
    21 -> 33 [ label = "" ]
    33 -> 14 [ label = "" ]
    20 -> 14 [ label = "" ]
}

